rfcs/convergence/no-css-shorthands-in-make-styles.md

```diff
@@ -0,0 +1,218 @@
+// ⚠ discarded solution, used only for an example
+makeStyles({
+  rootA: {
+    // 1️⃣ no custom properties
+    paddingX: 'var(--foo)',
+  },
+  rootB: {
+    // 2️⃣ no custom values
+    padding: { left: 'var(--foo)', right: 'var(--foo)' },
+    margin: { all: 'var(--foo)' },
+  },
+  rootC: {
+    // 2️⃣ no custom values
+    padding: ['var(--foo)', 'var(--foo)'],
+  },
+});
+```
+
+## Detailed Design or Proposal
+
+### Ban CSS shorthands in typings
+
+The first part of proposal is to ban CSS shorthands in typings: no shorthand properties = no need to expand them = no issues ✅
+
+Proposed to forbid usage of all CSS shorthands ([list of CSS shorthands exported from MDN](https://csstree.github.io/docs/syntax/#report&noedit&title=CSS%20shorthand%20properties%20by%20MDN%20data&q=JGlzQXJyYXk6ID0%2BICQgIT0gKCQgKyAnJyk7Cm1kbi5wcm9wZXJ0aWVzLmVudHJpZXMoKS4oeyBpZHgsIG5hbWU6IGtleSwgLi4uKHZhbHVlIHwgeyBjb21wdXRlZCwgc3ludGF4IH0pIH0pLltjb21wdXRlZC4kaXNBcnJheSgpXQ%3D%3D&v=ewogICAgdmlldzogJ3RhYmxlJywKICAgIGNvbHM6IHsKICAgICAgICBpZHg6ICd0ZXh0OiMuaW5kZXggKyAxJywKICAgICAgICBjb21wdXRlZDogewogICAgICAgICAgICBoZWFkZXI6ICdwcm9wcycsCiAgICAgICAgICAgIGNvbnRlbnQ6ICd1bDpjb21wdXRlZCcKICAgICAgICB9CiAgICB9LAogICAgbGltaXQ6IGZhbHNlCn0%3D)) except vendor prefixed (i.e. `-ms`, `-webkit`).
+
+```js
+makeStyles({
+  // 🚨 TypeScript will throw
+  rootA: { padding: '4px' },
+  // ✅ TypeScript will pass
+  rootB: { paddingLeft: '4px', paddingRight: '4px', paddingBottom: '4px', paddingTop: '4px' },
+});
+```
+
+### Create macro functions for expansion
+
+As it will be annoying and redundant to use CSS longhands in every case we will have special functions for expansion.
+
+```js
+import { macros } from '@fluentui/react-make-styles';
+
+makeStyles({
+  rootA: { paddingLeft: '4px', paddingRight: '4px', paddingBottom: '4px', paddingTop: '4px' },
+  rootB: {
+    ...macros.padding('4px'),
+    ...macros.padding('4px', '8px'),
+    ...macros.padding('4px', '8px', '16px', 0),
+  },
+});
+```
+
+`macros.padding('4px')` from the example above will expand it two four properties:
+
+```js
+shallowEqual(
+  { paddingLeft: '4px', paddingRight: '4px', paddingBottom: '4px', paddingTop: '4px' },
+  macros.padding('4px'),
+); // returns "true"
+```
+
+Each macro will implement arguments order & expand based on CSS spec.
+
+For initial implementation following macros will be implemented:
+
+- `border`
+- `borderLeft`
+- `borderBottom`
+- `borderRight`
+- `borderTop`
+- `borderColor`
+- `borderStyle`
+- `borderRadius`
+- `borderWidth`
+- `margin`
+- `padding`
+
+Additional functions can be implemented based on customer requests.
+
+- These functions will not have significant impact on runtime version of `makeStyles()` as they are very lightweight.
+- These functions will not **any** impact when Babel plugin/Webpack loader are used as they will be evaluated during build.
+
+### Pros and Cons
+
+- 👍 No CSS shorthands => no problems
+- 👍 No new syntax
+- 👎 Non obvious TypeScript errors in some cases
+
```

Comment:
I would add a con that we have to explain this to users of makeStyles in our documentation.  It causes Fluent to seem a little less aligned with web standards since people would expect to use shorthand with raw CSS and with most frameworks.

Reply:
Yeah, true, added 👍

Reply:
I'm afraid I don't have the full context so please take this with a grain of salt, but it doesn't make sense to me that complicating DX should be justified with adding documentation. As a user, I would expect a styling library to allow me to write CSS like I normally would.
Could we maybe contribute to `inline-style-expand-shorthand` to add the support we need instead or is this what you meant in the issue that it's not feasible in terms of complexity?

Reply:
@andrefcdias I agree, but during offline discussion we agreed to make something always predictable instead of _sometimes working_. 

---

The problem is covered in the summary section of this RFC and in the original issue #19402. There are different cases: in some it's _about complexity_, but there is the main case when it's _about possibility_ at all:

```ts
{ padding: 'var(--some-var)' }
```

👆 CSS variable can be anything (🗃️ [CodeSandbox](https://codesandbox.io/s/inline-style-expand-shorthand-css-variables-n9mh3)), it means that it's **unsafe** to expand it at all.

To summarize:
- CSS variables are not expandable at all during build time (we don't have CSS variables in Node environment even if _we would like to resolve and expand_)
- CSS can be resolved in runtime via calls of `getPropertyValue()` (🗃️ [CodeSandbox](https://codesandbox.io/s/practical-nightingale-85h08?file=/src/App.js)), but will be _too expensive_  as it should be done for every component and every usage of `theme` or CSS variables
  - Note 1: `getPropertyValue()` should be called **after** render as during we may not have DOM at all (👋 double rendering)
  - Note 2: we need to handle updates (How at all?)

The same problem is true for **any** CSS-in-JS engine that uses Atomic CSS (Compiled, Stylex, Fela, etc.).

Reply:
Thanks for clarifying! Just wanted to make sure we covered all our options.